### PR TITLE
requests: support sub-classed sessions

### PR DESCRIPTION
**Pull request by owais**
We patch the Session object such that it actually returns instances
of our subclass (SessionTracing). This breaks when users use a sub-class
of requests.Session as our implementation subclasses requests.Session
directly.

For example, when using requests-oauthlib, none of the methods provided
by requests_oauthlib.Session can be accessed as we actually force create
SessionTracing instances insted of requests_oauthlib.Session.

This commmit checks whether the user is requesting an instance of
requests.Session or a subclass of it. If it requests, it preserves the
current behavior. If it is a subclass, it dynamically creates another
class that derives from the subclass and SessionTracing.

### CODE REVIEW

File: signalfx_tracing/libraries/requests_/instrument.py [hunk 1]
```diff
@@ -18,10 +18,22 @@
...
-    return SessionTracing.__new__(SessionTracing)
+    from requests.sessions import Session
+    if session_class is Session:
+        return SessionTracing.__new__(SessionTracing)
+
+    # if a subclass of requests.Session is being used, create
+    # a class that inherits from the subclass and SessionTracing.
+    # This ensures the resulting session object is traced while
+    # still satisying the subclass being used.
+    class CustomSessionTracing(session_class, SessionTracing):
```
Comment (rmfitzpatrick) [inline signalfx_tracing/libraries/requests_/instrument.py:32]: Would we want `(SessionTracing, session_class)` to trace any overridden `request()` method?

File: signalfx_tracing/libraries/requests_/instrument.py [hunk 2]
```diff
@@ -18,10 +18,22 @@
...
 
 
-def session_new(_, __, *args, **kwargs):
+def session_new(_, session_class, *args, **kwargs):
     """Monkey patch Session.__new__() to create a SessionTracing object"""
     from requests_opentracing import SessionTracing
-    return SessionTracing.__new__(SessionTracing)
+    from requests.sessions import Session
+    if session_class is Session:
+        return SessionTracing.__new__(SessionTracing)
```
Comment (rmfitzpatrick) [inline signalfx_tracing/libraries/requests_/instrument.py:26]: Is this check necessary?

File: signalfx_tracing/libraries/requests_/instrument.py (hunk 1 above)
Comment (owais) [reply signalfx_tracing/libraries/requests_/instrument.py:32]: Python figures out the mro correctly in this case (diamond dep). Updated the test case to add a custom request method. As long as the custom method calls super correctly, it should get the custom method and tracing both.

File: signalfx_tracing/libraries/requests_/instrument.py (hunk 2 above)
Comment (owais) [reply signalfx_tracing/libraries/requests_/instrument.py:26]: Not really but felt weird to have yet another class subclassing from requests.Session and SessionTracing where SessionTracing already subclasses requests.Session.

File: signalfx_tracing/libraries/requests_/instrument.py (hunk 1 above)
Comment (owais) [reply signalfx_tracing/libraries/requests_/instrument.py:32]:
```
class Session:
    pass

class CustomSession(Session):
    pass

class TracedSession(Session):
     pass

class FinalCustomPlusTraced(CustomSession, TracedSession):
    pass

FinalCustomPlusTraced.__mro__
(__main__.FinalCustomPlusTraced, __main__.CustomSession, __main__.TracedSession, __main__.Session, object)
```

Comment (rmfitzpatrick) [reply signalfx_tracing/libraries/requests_/instrument.py:32]: My assumption would be that CustomSession's request doesn't call its super's request, which is why TracedSession would need to come before it in mro.

Comment (owais) [reply signalfx_tracing/libraries/requests_/instrument.py]: Updated so both cases are supported now.

**Review by rmfitzpatrick — APPROVED** [review]
Thanks!

**Pull request merged**